url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/math/MathExpression.java

```diff
@@ -0,0 +1,615 @@
+
+			if (hs.size() == 1) {
+
+				for (final MathsItem it : hs) {
+
+					if (it instanceof MathExpression) {
+
+						final MathExpression h = (MathExpression) it;
+						
+						for (final HashSet<MathsItem> hs2 : h.items) {
+							final HashSet<MathsItem> nmult = new HashSet<MathsItem>();
+							for (final MathsItem mit : hs2)
+								nmult.add(mit.clone());
+							items.add(nmult);
+						}
+						happend = true;
+						
+						continue outer;
+					} 
+
+					break;
+				}
+
+			}
+
+			items.add(hs);
+
+		}
+		
+		return happend;
+
+	}
+	
+	private boolean performAdditionOnSimpleExpressions() {
+		
+		final HashSet<HashSet<MathsItem>> toremove = new HashSet<HashSet<MathsItem>>();
+		boolean happened = true;
+		boolean h = false;
+		
+		while(happened) {
+			
+			toremove.clear();
+			happened = false;
+			
+			outer : for (final HashSet<MathsItem> m1 : items)
+				if (m1.size() == 1)
+					for (final HashSet<MathsItem> m2 : items)
+						if (m1 != m2 && m2.size() == 1) {
+							
+							oneloop : for (final MathsItem i1 : m1)
+								for (final MathsItem i2 : m2) {
+									
+									if (i1.add(i2)) {
+										
+										happened = true;
+										toremove.add(m2);
+										break outer;
+										
+									}
+									
+									break oneloop;
+								}
+							
+						}
+			
+			for (final HashSet<MathsItem> m : toremove)
+				items.remove(m);
+			
+			h |= !toremove.isEmpty();
+			
+		}
+		
+		return h;
+	}
+
+	private boolean removeAllOnes() {
+		boolean happened = false;
+		
+		for (final HashSet<MathsItem> hs : items) {
+			
+			if (hs.size() < 2)
+				continue;
+			
+			final HashSet<MathsItem> toremove = new HashSet<MathsItem>();
+
+			
+			for (final MathsItem m : hs)
+				if (m.isOne())
+					toremove.add(m);
+			
+			if (toremove.size() == hs.size()) {
+				int i = 0;
+				for (final MathsItem m : toremove) {
+					if (i != 0) hs.remove(m);
+					i++;
+				}
+				
+				if (i <= 1)
+					continue;
+			} else {
+				for (final MathsItem m : toremove)
+					hs.remove(m);
+			}
+			
+			happened |= !toremove.isEmpty();
+		}
+		
+		return happened;
+	}
+	
+	private boolean removeAllZeros() {
+		
+		final HashSet<HashSet<MathsItem>> toremove = new HashSet<HashSet<MathsItem>>();
+		
+		for (final HashSet<MathsItem> hs : items) {
+
+			for (final MathsItem m : hs)
+				if (m.isZero()) {
+					toremove.add(hs);
+					continue;
+				}
+		}
+		
+		if (toremove.size() == items.size()) {
+			int i = 0;
+			
+			for (final HashSet<MathsItem> m : toremove) {
+				if (i != 0)
+					items.remove(m);
+				else {
+					// if we are to leave a zero element, put it to zero
+					m.clear();
+					m.add(new MathNumber(0));
+				}
+				i++;
+			}
+			
+			if (i <= 1)
+				return false;
+			
+		} else {
+			for (final HashSet<MathsItem> m : toremove)
+				items.remove(m);
+		}
+		
+		return !toremove.isEmpty();
+	}
+	
+	/**
+	 * Translates a*3 + a*5 into a*8
+	 * @return
+	 */
+	private boolean commonFactors() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.